### PR TITLE
Changes to definitions for NXmx Gold Standard.  Includes changes to NXdl

### PR DESCRIPTION
**Pull request by yayahjb**
This is the new NXmx Gold Standard as discussed at the raw data e-workshop, 22 August 2020 for
consideration by NIAC.  There is a companion branch for cnxvalidate -- HJB

Includes a fix for #782

### CODE REVIEW

**Comment by vasole** [comment]
It had been approved to use DATASET_errors to denote the uncertainties associated to DATASET. However, in this PR it uses DATASET_error (see flatfield). Is the missing s a typing mistake, or an exception to the rule that has to be considered?

**Comment by yayahjb** [comment]
Good catch

It appears an inconsistency in the use of _error versus _errors has crept
into some definitions, with some
uses of _errors and some uses of _error.  I think we should pick one or the
other as the "correct" choice going
forward, but special case whichever of the two then becomes deprecated to
say that software should accept the
deprecated version as equivalent at least for some given period.  Here are
some of the use cases in
current use:

  variable_errors
 data_error
 flatfield_error
 average_value_error
 VARIABLE_errors
 wavelength_error
 energy_error

taken from definitions.  There is no way to know about other uses in hdf5
datasets.
When we meet on 4 September we should resolve this one way or the other in
terms of
NXDL.  In the meantime, I'll draft some patch code for cnxvalidate to warn
about
use of, say, _error

  -- Herbert

**Comment by vasole** [comment]
In the NeXus documentation it was always writen in plural and therefore adopted in latest NIAC as the preferred form.

In the application definitions might have been something else.

https://github.com/nexusformat/definitions/issues/634

**Comment by vasole** [comment]
LGTM.

**Comment by yayahjb** [comment]
To all:  As discussed in the Telco, please complete your review of this
pull request by 12 noon NY time on Wednesday 9 Sep 2020.

File: nxdl.xsd [hunk 1]
```diff
@@ -608,23 +623,19 @@
 				<xs:sequence>
 					<xs:element name="dimensions" type="nx:dimensionsType" minOccurs="0" maxOccurs="1">
 						<xs:annotation>
-							<xs:documentation>
-								dimensions of a data element in a NeXus file
-							</xs:documentation>
+							<xs:documentation>dimensions of a data element
```
Comment (PeterC-DLS) [inline nxdl.xsd:626]: Needless formatting

File: applications/NXmx.nxdl.xml [hunk 2]
```diff
@@ -80,77 +107,149 @@
...
+						/Items/_diffrn_source.type.html} is highly recommended.
+					</doc>
+					<attribute name="short_name">
+						<doc>Short name for instrument, perhaps the acronym.</doc>
+					</attribute>
+				</field>
+
+				<field name="time_zone" type="NX_DATE_TIME" recommended="true">
+					<doc>
+						ISO 8601 time_zone offset from UTC.
```
Comment (PeterC-DLS) [inline applications/NXmx.nxdl.xml:187]: Why is this needed if times above are recorded in UTC?

File: applications/NXmx.nxdl.xml [hunk 3]
```diff
@@ -267,104 +371,115 @@
...
+							Normally this value is for guidance only, the proper 
+							geometry can be found following the depends_on axis chain,
+							But in appropriate cases where the dectector distance
+							to the sample is observable independent of the axis
+							chain, that may take precedence over the axis chain
+							calculation.  
+						</doc>
+					</field>
+
+					<field name="distance_derived" type="NX_BOOLEAN" units="NX_LENGTH" recommended="true">
```
Comment (PeterC-DLS) [inline applications/NXmx.nxdl.xml:431]: Units here is incorrect

File: applications/NXmx.nxdl.xml [hunk 4]
```diff
@@ -267,104 +371,115 @@
...
+					</field>
+
+					<field name="count_time" type="NX_NUMBER" units="NX_TIME" recommended="true">
+						<doc>
+							Elapsed actual counting time.
 						</doc>
 					</field>
 
-					<field name="count_time" type="NX_NUMBER" units="NX_TIME" minOccurs="0">
+					<field name="beam_center_derived" type="NX_BOOLEAN" units="NX_LENGTH" optional="true">
```
Comment (PeterC-DLS) [inline applications/NXmx.nxdl.xml:452]: Units attribute incorrect

File: applications/NXmx.nxdl.xml [hunk 5]
```diff
@@ -80,77 +107,149 @@
...
+						Use of the depends_on field and the NXtransformations group is
+						strongly recommended.  As noted above this should be an absolute
+						requirement to have for any scan experiment.
+
+						The reason it is optional is mainly to accommodate XFEL
+						single shot exposures.
+					</doc>
+				</group>
+
+				<field name="temperature" units="NX_TEMPERATURE" minOccurs="0" />
```
Comment (PeterC-DLS) [inline applications/NXmx.nxdl.xml:165]: Missing type

File: applications/NXmx.nxdl.xml [hunk 6]
```diff
@@ -393,16 +508,34 @@
 					</field>
 
 					<field name="flatfield" type="NX_FLOAT" minOccurs="0">
-						<doc>Flat field correction data.</doc>
+						<doc>
+							Flat field correction data.  If provided, it is recommended
+							that is be compressed.
```
Comment (PeterC-DLS) [inline applications/NXmx.nxdl.xml:513]: Typo it not is

File: applications/NXmx.nxdl.xml [hunk 7]
```diff
@@ -393,16 +508,34 @@
 					</field>
 
 					<field name="flatfield" type="NX_FLOAT" minOccurs="0">
```
Comment (PeterC-DLS) [inline applications/NXmx.nxdl.xml:510]: Should type be number here and below?

File: applications/NXmx.nxdl.xml [hunk 8]
```diff
@@ -413,42 +546,38 @@
 					<field name="pixel_mask_applied" type="NX_BOOLEAN" minOccurs="0">
 						<doc>
 							True when the pixel mask correction has been applied in the
-							electronics, false otherwise.
+							electronics, false otherwise.  
 						</doc>
 					</field>
 
-					<field name="pixel_mask" type="NX_INT" minOccurs="0">
+					<field name="pixel_mask" type="NX_INT" recommended="true">
```
Comment (PeterC-DLS) [inline applications/NXmx.nxdl.xml:553]: Doc implies type should be NX_INT32

**Review by PeterC-DLS — CHANGES_REQUESTED** [review]

File: applications/NXmx.nxdl.xml [hunk 9]
```diff
@@ -80,77 +107,149 @@
...
+
+
 			<group type="NXinstrument">
-			        <field name="name" minOccurs="1">
-                                       <doc>Name of instrument</doc>
-                                       <attribute name="short_name">
-                                               <doc>short name for instrument, perhaps the acronym</doc>
-                                       </attribute>
-                                </field>
+				<field name="name" minOccurs="1">
```
Comment (benajamin) [inline applications/NXmx.nxdl.xml:171]: Shouldn't this  also have type="NXchar"?

**Comment by prjemian** [comment]
Not necessary to specify. NXchar is the default.

File: applications/NXmx.nxdl.xml [hunk 10]
```diff
@@ -267,104 +371,115 @@
 									<item value="translation" />
 								</enumeration>
 							</attribute>
-							<attribute name="vector">
+							<attribute name="vector" type="NX_NUMBER">
 							</attribute>
-							<attribute name="offset">
+							<attribute name="offset" type="NX_NUMBER">
```
Comment (benajamin) [inline applications/NXmx.nxdl.xml:376]: I am confused here, perhaps I am missing something. I understand that the "module_offset" field has the direction and magnitude given by the "vector" and "offset" attribute, respectively. However, shouldn't a field have a value associated with it? If all the relevant info is given by attributes, then what is left to write in the field value?
The same issue is repeated a few times throughout the NXdetector_module base class, so it isn't just an NXmx issue.

**Review by benajamin — APPROVED** [review]

File: applications/NXmx.nxdl.xml (hunk 10 above)
Comment (PeterC-DLS) [reply applications/NXmx.nxdl.xml:376]: It's specified [in](https://manual.nexusformat.org/design.html#coordinate-transformation-field-and-attributes) for all NXtransformations fields.

File: nxdl.xsd [hunk 11]
```diff
@@ -792,12 +818,12 @@
 				<xs:attribute name="data_offset" use="optional"  default="1" type="nx:nonNegativeUnbounded">
 					<xs:annotation>
 						<!--
-							note: renamed from "offset" due to conflict with CIF usage.
+							note: renamed from "offset" due to comflict with CIF usage.
```
[the remark this answers is not in the document]
Comment (yayahjb) [reply nxdl.xsd]: Fixed by commit 1441160

File: nxdl.xsd (hunk 1 above)
Comment (yayahjb) [reply nxdl.xsd]: Fixed by commit  69b555e

File: applications/NXmx.nxdl.xml (hunk 3 above)
Comment (yayahjb) [reply applications/NXmx.nxdl.xml]: Fixed commit 50f7c00

File: applications/NXmx.nxdl.xml (hunk 6 above)
Comment (yayahjb) [reply applications/NXmx.nxdl.xml]: Fixed commit 50f7c00

File: applications/NXmx.nxdl.xml (hunk 7 above)
Comment (yayahjb) [reply applications/NXmx.nxdl.xml]: Fixed commit 50f7c00

File: applications/NXmx.nxdl.xml (hunk 8 above)
Comment (yayahjb) [reply applications/NXmx.nxdl.xml:553]: I think NX_INT is adequate, but if somebody feels strongly they should change ii

File: applications/NXmx.nxdl.xml (hunk 5 above)
Comment (yayahjb) [reply applications/NXmx.nxdl.xml]: Fixed commit 50f7c00

File: applications/NXmx.nxdl.xml (hunk 2 above)
Comment (yayahjb) [reply applications/NXmx.nxdl.xml:187]: For use in resolving issues related to other time dependent information

File: applications/NXmx.nxdl.xml (hunk 8 above)
Comment (PeterC-DLS) [reply applications/NXmx.nxdl.xml:553]: So if bit depth < 32 implies missing fields are false/unset?

**Comment by yayahjb** [comment]
yes

**Review by PeterC-DLS — APPROVED** [review]

File: applications/NXmx.nxdl.xml [hunk 12]
```diff
@@ -652,11 +744,6 @@
 						<doc>
 							The wavelength spread FWHM for the corresponding
 							wavelength(s) in incident_wavelength.
-
```
Comment (phyy-nx) [inline applications/NXmx.nxdl.xml:655]: Likewise, could this note be preserved?  Added in #717 as part of handling per-shot wavelengths.

**Review by phyy-nx — CHANGES_REQUESTED** [review]
Other than requesting these two notes be restored, changes look good to me.

File: applications/NXmx.nxdl.xml (hunk 9 above)
Comment (yayahjb) [reply applications/NXmx.nxdl.xml]: Fixed in commit  1fe10e2

File: applications/NXmx.nxdl.xml [hunk 13]
```diff
@@ -625,12 +723,6 @@
 							of dimensions **np** by **m** (slow to fast) with the
 							relative weights in incident_wavelength_weight as a 2D
 							array.
-
```
[the remark this answers is not in the document]
Comment (yayahjb) [reply applications/NXmx.nxdl.xml:628]: Fixed in commit  1fe10e2

File: applications/NXmx.nxdl.xml (hunk 12 above)
Comment (yayahjb) [reply applications/NXmx.nxdl.xml:655]: Fixed in commit  1fe10e2

File: applications/NXmx.nxdl.xml (hunk 10 above)
Comment (yayahjb) [reply applications/NXmx.nxdl.xml:376]: I don't know.  My guess would be that, as with axis setting, we could give a value (or values) that would then have the effect of giving a displacement from the offset in the
direction of the vector, but if there actually is a dynamic positioner it would be better
to explicitly describe the relevant positioner axis and its settings.  For the moment
I will go forward with the current pull request and suggest we discuss this point further in the future, perhaps at NIAC.

**Comment by yayahjb** [comment]
Thanks to all who have commented.  Unless i misunderstand, I think everybody is comfortable with merging the changes we now have at noon NY time today, but we have discovered some interesting old/new things to discuss at NIAC

**Review by phyy-nx — APPROVED** [review]

**Comment by yayahjb** [comment]
Merged as proposed -- Herbert

**Comment by prjemian** [comment]
added to next milestone so this PR will be recorded automatically in the release notes
